utils/gui.py

```diff
@@ -30,7 +30,10 @@ def windowToFront(root):
         pythoncom.CoInitialize()
         shell = win32com.client.Dispatch("WScript.Shell")
         shell.SendKeys("%")
-        win32gui.SetForegroundWindow(root.winfo_id())
+        try: # Fails on mock test on windows
+            win32gui.SetForegroundWindow(root.winfo_id())
+        except Exception as e:
+            pass
```

Comment:
Silencing this by catching it and passing does not tell us why it's failing, it just silences any information we get about the real problem.

Reply:
It fails if the window no longer exist, or in the case of mock testing we dont have a windows id.
I dont think theres any point in handling it in any special case since theres nothing to do about it

Reply:
Blanket catching an "Exception" basically catches any exception thrown in that path. We're silencing any possibility of anything else going wrong by doing this.

Reply:
But thats the only thing that does go wrong with this, what would you have it do?

Reply:
Currently, from your perspective, but it assumes a lot about the future or the external dependency in general. If you have proof that this is the only way needed to handle any exception path through that function, then include it in the PR.

Reply:
The solution here would be to mock out that function so that it becomes a no-op. It hasn't been a problem because Travis only runs the tests through Linux and Darwin (Mac).

Reply:
It might also crash the program if a window dissapears for other reasons

Reply:
In that case, we need to be handling the exceptions that are being thrown for other reasons. We may need to mitigate different causes; blanket passing all exceptions does not help that. It will simply silently fail without any information.

Reply:
But why not let it silenty fail? If the window does not exist anymore theres not much to do about it. Later on if we have some kind of debug log we can have it write out there.

Reply:
No. If this causes a bug during runtime currently, we need it to crash, as it should, and be reported. If we aren't handling a case properly, that's an issue in the code. Silently failing does not solve that problem.

Reply:
Why should the entire program crash because it couldnt refocus a window that does not exist anymore? It will just confuse the end users. And with gui changes we might not need this function at all anymore

Reply:
Because if we're not dealing with that properly, it should be fixed so that it does deal with it properly. Again, silently failing does nothing to help that issue. It's simply prolonging issues in the codebase without dealing with them.

Reply:
So I ask again, how would you deal with it properly?

Reply:
For now, if you want tests to not cause that problem, you should be mocking out those functions. If we can reproduce that issue while running the program in production, that should be reported as a Github Issue, with steps to reproduce. If we have that information about actual use in production, we can begin to solve the problem. For testing, we are not testing that external dependencies work. We are testing that our internal code paths are not causing unexpected problems.

Reply:
That being said, having an actual test that could test the usage of external dependencies would be great, but that's some work that should be separated from the mocked GUI testing we have.

Reply:
> For now, if you want tests to not cause that problem, you should be mocking out those functions. If we can reproduce that issue while running the program in production, that should be reported as a Github Issue, with steps to reproduce. If we have that information about actual use in production, we can begin to solve the problem. For testing, we are not testing that external dependencies work. We are testing that our internal code paths are not causing unexpected problems.

To clarify here; obviously, if you can completely get an understanding of the problem and introduce a PR that solves any bug that actually exists in production, the Issue is not needed... That's just a standard path that should be taken.

Reply:
I still dont think we should crash the program over something that could be ignored, I can make it print a message to console for now

Reply:
Alright, good luck.

Reply:
?

Reply:
Good luck with your PR.